### ISSUE TITLE
Branch processor with dynamic args in result_map

### ISSUE THREAD

**Issue by mihaitodor** [Jeffail/benthos#633]
I was toying around with a non-trivial process pipeline and I bumped into an issue with using bloblang in a `result_map`. I dug a bit through the code, but I'm not yet sure if I hit a bug or if I'm doing something wrong.

Given this prices.csv input file:
```csv
Date,High,Low
"2018-02-20",965.26,892.95
"2018-02-19",957.78,921.55
"2018-02-18",982.93,915.45
"2018-02-17",976.60,940.75
"2018-02-16",950.00,917.85
"2018-02-15",951.95,911.11
```

and config yaml:

```yaml
input:
  type: file
  file:
    paths: [ "prices.csv" ]
    codec: csv
pipeline:
  processors:
    - try:
      - bloblang: |
          Date = this.Date
          AvgUSD = (this.High.number() + this.Low.number()) / 2
      - branch:
          request_map: 'root = this'
          processors:
            - http:
                url: 'https://api.exchangeratesapi.io/history?base=USD&start_at=${! json("Date") }&end_at=${! json("Date") }&symbols=EUR'
                verb: GET
                retries: 0
          result_map: |
            root.Rates = this.rates
      - bloblang: |
          Date = this.Date
          AvgUSD = this.AvgUSD
          AvgEUR = this.AvgUSD * this.Rates.get(this.Date).EUR.or(0)
output:
 type: stdout
 stdout:
   delimiter: "\n"
```

produces this output:

```
{"AvgEUR":752.9213938645075,"AvgUSD":929.105,"Date":"2018-02-20"}
{"AvgEUR":757.183722838112,"AvgUSD":939.665,"Date":"2018-02-19"}
{"AvgEUR":0,"AvgUSD":949.19,"Date":"2018-02-18"}
{"AvgEUR":0,"AvgUSD":958.675,"Date":"2018-02-17"}
{"AvgEUR":749.2979781906974,"AvgUSD":933.925,"Date":"2018-02-16"}
{"AvgEUR":745.6415592540301,"AvgUSD":931.53,"Date":"2018-02-15"}
```

while this config yaml:

```yaml
input:
  type: file
  file:
    paths: ["prices.csv"]
    codec: csv
pipeline:
  processors:
    - try:
        - bloblang: |
            Date = this.Date
            AvgUSD = (this.High.number() + this.Low.number()) / 2
        - branch:
            request_map: "root = this"
            processors:
              - http:
                  url: 'https://api.exchangeratesapi.io/history?base=USD&start_at=${! json("Date") }&end_at=${! json("Date") }&symbols=EUR'
                  verb: GET
                  retries: 0
            result_map: |
              root.EURPrice = this.rates.get(this.Date).EUR
        - bloblang: |
            Date = this.Date
            AvgUSD = this.AvgUSD
            AvgEUR = this.AvgUSD * this.EURPrice.or(0)
output:
  type: stdout
  stdout:
    delimiter: "\n"
```

produces this output:

```
{"AvgUSD":929.105,"Date":"2018-02-20"}
{"AvgUSD":939.665,"Date":"2018-02-19"}
{"AvgUSD":949.19,"Date":"2018-02-18"}
{"AvgUSD":958.675,"Date":"2018-02-17"}
{"AvgUSD":933.925,"Date":"2018-02-16"}
{"AvgUSD":931.53,"Date":"2018-02-15"}
```

From my investigation, it looks like gabs.Search is swallowing errors [here](https://github.com/Jeffail/gabs/blob/e47bd8079aa6dcc9e48a4035f3f1714b54dc2dcd/gabs.go#L187). In this case, [this one](https://github.com/Jeffail/gabs/blob/e47bd8079aa6dcc9e48a4035f3f1714b54dc2dcd/gabs.go#L144) gets ignored... Not yet sure why this results in the `AvgEUR` field getting silently dropped from the output, though.

Trying to figure out why this error is produced in the first place, it seems to consider `this` as the document returned by that HTTP call. I guess calling `get(this.Date)` in the `result_map` of the branch processor is not supported. What is the right way to reference the `Date` field from the previous processor in the `result_map`?

The reason for sticking a `branch` in there was to get the `Date` field to be propagated to the final output and the `request_map: 'root = this'` part is probably unnecessary in there, because that's the default behaviour, if I understand the docs correctly.

~~Also, if you notice in the correct output, there are 2 entries where `"AvgEUR":0`. I'd like to remove these rows somehow via bloblang's `delete()` function, but couldn't figure out how to achieve it.~~ This bit can be achieved by adding another `bloblang` processor at the end of the pipeline:
```yaml
- bloblang: |
    root = match {
      this.AvgEUR == 0 => deleted()
      _ => this
    }
```

**Comment by Jeffail** [comment]
Hey @mihaitodor, it looks like a mapping error might be occurring that results in the third mapping not being attempted, my guess would be that in the result map `root.AvgEUR = this.rates.get(this.Date)` the field `this.Date` isn't present in the response from the http request, and therefore an error will be thrown as `this.Date` becomes `null` which isn't valid as an argument to the `get` method. If you enable debug logging you should see the errors as they occur, or add a `catch` block with a log in it.

**Comment by mihaitodor** [comment]
Hey @Jeffail, thank you for the reply! Answers below:

> the field `this.Date` isn't present in the response from the http request

Correct, because it comes from the input of this `branch` processor. I guess my problem is how can I get values from fields coming from the `request_map` in the `result_map`. Is this some sort of anti-pattern?

> If you enable debug logging you should see the errors as they occur,

Indeed, that reveals `{"@service":"benthos","level":"DEBUG","component":"benthos.pipeline.processor.0.1","message":"Failed to map result '0': failed to execute mapping query at line 1: expected string argument, received <nil>"}` as expected.

So, basically, is my first yaml the way to solve this problem? As in let the `branch` processor send the whole payload to the last `bloblang` processor and then have that extract what it needs from the payload (i.e. the field under the key with value `this.Date`). Intuitively, I was expecting the second yaml to work.

**Comment by Jeffail** [comment]
You could use metadata to carry the date through like this:

```yaml
pipeline:
  processors:
    - try:
      - bloblang: |
          Date = this.Date
          AvgUSD = (this.High.number() + this.Low.number()) / 2
      - branch:
          request_map: |
            root = this
            meta date = this.Date
          processors:
            - http:
                url: 'https://api.exchangeratesapi.io/history?base=USD&start_at=${! json("Date") }&end_at=${! json("Date") }&symbols=EUR'
                verb: GET
                retries: 0
          result_map: 'root.AvgEUR = this.rates.get(meta("date"))'
      - bloblang: |
          Date = this.Date
          AvgUSD = this.AvgUSD
          AvgEUR = this.EUR.or(0)
```

But your first example is closer to the general intention of the branch processor, where you use a branch to compose the nested processors, and then follow up the result map with another bloblang processor if you need to perform a merge of the result with the original data.

**Comment by mihaitodor** [comment]
@Jeffail Thank you so much for that explanation! Indeed, passing it via the `meta` didn't occur to me! However, it does feel a bit awkward to do it that way. I just noticed that my second example was wrong and confusing, so I updated it.

Intuitively, I'd expect the `branch` processor to not expose more information than it needs to for the next step, which is why I wasn't happy with my first approach and thought there has to be a straightforward way to reference the input fields of the `branch` processor in the `result_map`. Not sure what would be the semantically correct way to reference them, though, if that's something you'd be interested in supporting.

Anyhow, I'm OK with the provided workaround, so please feel free to close the issue.
